docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
+- Performs Helm uninstall on `HelmRelease` CR removal
+- Chart values override\
+- Pull charts from Helm repository directly
+
+_Develop our own Homebrew Helm Controller_
```

Comment:
I am not sure if I can go along with because we will need to create the controllers to manage the AddOns.
The Helm part that we need is only to manage helm charts itself (install/uninstall/upgrade) all the other needs still in place indeed if we use Flux.

Reply:
It looks to me like helm does not update crds

https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

and thus flux has added logic and policies around crds

https://github.com/fluxcd/helm-controller/blob/main/internal/runner/runner.go#L229C31-L230

I cant say what other complexities we will find when trying to implement this ourselves and thus it may be best to offload this to flux

Reply:
Additionally flux has built in test support for validating releases. This could be beneficial

https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go#L436

Reply:
@emosbaugh, 

Flux is who manages and that is the whole issue. 
It means we are NOT allow to prevent the solution be uninstalled since we are NOT responsible for manage it. 
Therefore, see that HelmController is generic and flux never will allow us to do our specific checks prior allow or not solution get uninstalled. Example: is all PVC migrates? Have you other object store installed and etc.

Reply:
In these comments I am only listing the benefits of using flux. I understand you point about uninstalling.

Reply:
> Therefore, see that HelmController is generic and flux never will allow us to do our specific checks prior allow or not solution get uninstalled. Example: is all PVC migrates? Have you other object store installed and etc.

I don't understand why we can't perform checks like PVC migrations and object store migrations before uninstalling via Flux. Flux won't remove the object store until we tell it too, why can't we do our migrations, verifications, and then tell flux to remove it when we're ready?

Reply:
@chris-sanders @emosbaugh @ricardomaraschini @rrpolanco @laverya:

The primary issue with this approach, as per SOLID principles, particularly the Single Responsibility Principle (SRP) and the Liskov Substitution Principle (LSP), is that each class (or in this case, operator) should have one, and only one, reason to change, and that subclasses (or dependent services) must be substitutable for their base classes without altering the correctness of the program.

If we were to start manipulating Persistent Volume Claims (PVCs), or any resources managed by Flux as part of its reconciliation loop, it essentially means that our custom operator would be taking on more than one responsibility. This would violate the SRP, as our operator should only be concerned with its unique tasks, while Flux is tasked with managing the Helm charts.

In addition, by performing operations on resources managed by Flux before Flux has completed its own tasks, we're violating the LSP. Flux and our operator can be seen as base and derived classes, respectively. By interfering with Flux's responsibilities, we risk introducing errors into the system, because our operator would no longer be a drop-in replacement for Flux.

In a real-world scenario, imagine a race where both operators want to update or modify the PVCs. If we migrate and validate the PVCs independently and then allow Flux to remove them, there's a chance Flux could overwrite our changes in its reconciliation loop. This can lead to inconsistencies in the state of resources, potentially causing conflicts, data loss, or system instability. In essence, any changes we make could be nullified by Flux, wasting resources and leading to an endless cycle of modifications. Therefore, it is critical to respect each operator's boundaries and allow them to manage their designated resources independently.